File: ios/YamabikoChat/Features/Fusion/FusionOrchestrator.swift
import Foundation

struct FusionOrchestrator: Sendable {
    typealias Invoke = @Sendable (ProviderRequest, String, FusionPhase, @escaping @Sendable (ToolActivityPayload) -> Void) async throws -> ProviderResponse
    typealias CostEstimator = @Sendable (String, String, ProviderUsage?) async -> Double?
    typealias RequestBuilder = @Sendable (
        PanelModelConfig,
        String,
        FusionPhase,
        Bool
    ) async throws -> ProviderRequest

    typealias ProgressHandler = @Sendable (FusionProgressSnapshot) -> Void

    func runThroughJudge(
        request: FusionRequest,
        context: FusionContext,
        buildRequest: @escaping RequestBuilder,
        invoke: @escaping Invoke,
        estimateCost: @escaping CostEstimator,
        onProgress: ProgressHandler? = nil
    ) async throws -> FusionJudgeOutcome {
        let requestId = UUID().uuidString
        let startedAtMs = Int64(Date().timeIntervalSince1970 * 1000)

        if context.fusionDepth >= FusionContext.maxFusionDepth {
            throw FusionError.recursionLimitExceeded
        }

        let panelResults = await FusionPanelRunner.runAll(
            request: request,
            panelSystemPrompt: FusionPrompts.panelSystemPrompt(),
            buildPanelRequest: { panel, systemPrompt in
                try await buildRequest(panel, systemPrompt, .panel, request.allowWebSearch)
            },
            invoke: invoke,
            estimateCost: estimateCost,
            onProgress: onProgress
        )

        let successfulPanels = panelResults.filter { $0.success }
        let failedModels = panelResults.filter { !$0.success }.map { $0.modelId }

        guard !successfulPanels.isEmpty else {
            throw FusionError.allPanelsFailed(panelResults: panelResults)
        }

        let finalPanelChips = panelResults.map { result in
            FusionPanelChipStatus(
                modelId: result.modelId,
                provider: result.provider,
                state: result.success ? .succeeded : .failed
            )
        }
        onProgress?(
            FusionProgressSnapshot.phaseOnly(.judge, panels: finalPanelChips)
        )

        let judgeOutcome = try await runJudge(
            request: request,
            successfulPanels: successfulPanels,
            failedModels: failedModels,
            buildRequest: buildRequest,
            invoke: invoke,
            estimateCost: estimateCost
        )

        let judgeAnalysis = judgeOutcome.analysis
        let synthesisUserPrompt = judgeAnalysis != nil
            ? FusionPrompts.synthesizerUserPrompt(
                userPrompt: request.userPrompt,
                judgeAnalysis: judgeAnalysis,
                rawPanels: successfulPanels,
                judgeParseFailed: !judgeOutcome.parseSucceeded
            )
            : FusionPrompts.synthesizerUserPromptWithoutJudge(
                userPrompt: request.userPrompt,
                rawPanels: successfulPanels
            )

        let synthSystemPrompt = FusionPrompts.synthesizerSystemPrompt(debugMode: context.debugMode)
        var synthesisRequest = try await buildRequest(
            request.synthesizerModel,
            synthSystemPrompt,
            .synthesizer,
            false
        )
        synthesisRequest.messages = [
            ProviderRequestMessage(role: "user", content: synthesisUserPrompt)
        ]
        synthesisRequest.stream = true
        synthesisRequest.systemPrompt = SystemPromptComposer.mergeForAPI(
            request.systemPrompt,
            synthSystemPrompt
        )

        let trace = FusionTrace(
            requestId: requestId,
            preset: request.preset,
            startedAtMs: startedAtMs,
            completedAtMs: nil,
            panelResults: panelResults,
            judgeResult: judgeOutcome,
            synthesisResult: nil,
            totalLatencyMs: nil,
            totalCost: Self.sumCosts(panelResults: panelResults, judge: judgeOutcome, synthesis: nil),
            failedModels: failedModels,
            status: "judge_complete",
            userPrompt: context.logPrompts ? request.userPrompt : nil,
            finalAnswer: nil
        )

        let panelUsages = panelResults.compactMap { panel -> (String, String, ProviderUsage?, String)? in
            guard panel.success else { return nil }
            let usage = ProviderUsage(
                inputTokens: panel.inputTokens,
                outputTokens: panel.outputTokens
            ).normalizedNonEmpty()
            let requestType = "fusion_panel_\(Self.sanitizedModelId(panel.modelId))"
            return (panel.provider, panel.modelId, usage, requestType)
        }

        let judgeUsage: (String, String, ProviderUsage?)? = {
            guard judgeOutcome.parseSucceeded || judgeOutcome.rawJSON != nil else { return nil }
            let usage = ProviderUsage(
                inputTokens: judgeOutcome.inputTokens,
                outputTokens: judgeOutcome.outputTokens
            ).normalizedNonEmpty()
            return (request.judgeModel.provider.uppercased(), request.judgeModel.modelId, usage)
        }()

        onProgress?(
            FusionProgressSnapshot.phaseOnly(.synthesizer, panels: finalPanelChips)
        )

        return FusionJudgeOutcome(
            trace: trace,
            synthesisRequest: synthesisRequest,
            synthesizerProvider: request.synthesizerModel.provider,
            synthesizerModel: request.synthesizerModel,
            panelTokenUsages: panelUsages,
            judgeTokenUsage: judgeUsage
        )
    }

    func finalizeTrace(
        trace: FusionTrace,
        synthesisResult: SynthesisPhaseResult,
        finalAnswer: String,
        logPrompts: Bool
    ) -> FusionTrace {
        var updated = trace
        updated.synthesisResult = synthesisResult
        updated.completedAtMs = Int64(Date().timeIntervalSince1970 * 1000)
        updated.totalLatencyMs = (updated.completedAtMs ?? updated.startedAtMs) - updated.startedAtMs
        updated.totalCost = Self.sumCosts(
            panelResults: updated.panelResults,
            judge: updated.judgeResult,
            synthesis: synthesisResult
        )
        updated.status = "completed"
        updated.finalAnswer = logPrompts ? finalAnswer : nil
        return updated
    }

    // MARK: - Private

    private func runJudge(
        request: FusionRequest,
        successfulPanels: [PanelResult],
        failedModels: [String],
        buildRequest: @escaping RequestBuilder,
        invoke: @escaping Invoke,
        estimateCost: @escaping CostEstimator
    ) async throws -> JudgePhaseResult {
        let started = Date()
        let judgeUserContent = FusionPrompts.judgeUserPrompt(
            userPrompt: request.userPrompt,
            successfulPanels: successfulPanels,
            failedModels: failedModels
        )

        @Sendable func makeJudgeRequest(systemPrompt: String, userContent: String) async throws -> ProviderRequest {
            var providerRequest = try await buildRequest(
                request.judgeModel,
                systemPrompt,
                .judge,
                false
            )
            providerRequest.messages = [ProviderRequestMessage(role: "user", content: userContent)]
            providerRequest.stream = false
            providerRequest.metadata["responseMimeType"] = "application/json"
            if request.judgeModel.provider.uppercased() == "GEMINI" {
                providerRequest.metadata["geminiResponseMimeType"] = "application/json"
            }
            return providerRequest
        }

        let provider = request.judgeModel.provider

        let response = try await invoke(
                try await makeJudgeRequest(
                    systemPrompt: FusionPrompts.judgeSystemPrompt(),
                    userContent: judgeUserContent
                ),
                provider,
                .judge,
                { _ in }
            )
            let latencyMs = Int64(Date().timeIntervalSince(started) * 1000)
            let usage = response.usage?.normalizedNonEmpty()
            let cost = await estimateCost(request.judgeModel.provider, request.judgeModel.modelId, usage)

        if let analysis = FusionJudgeParser.parse(response.text) {
            return JudgePhaseResult(
                    analysis: analysis,
                    rawJSON: response.text,
                    parseSucceeded: true,
                    latencyMs: latencyMs,
                    inputTokens: usage?.inputTokens,
                    outputTokens: usage?.outputTokens,
                    cost: cost,
                    error: nil,
                    piExecutions: [response.piExecution].compactMap { $0 }
                )
        }

        let repairStarted = Date()
        let repairResponse = try await invoke(
                try await makeJudgeRequest(
                    systemPrompt: FusionPrompts.judgeSystemPrompt(),
                    userContent: FusionPrompts.jsonRepairPrompt(invalidJSON: response.text)
                ),
                provider,
                .judge,
                { _ in }
            )
            let repairLatency = Int64(Date().timeIntervalSince(repairStarted) * 1000)
            let repairUsage = repairResponse.usage?.normalizedNonEmpty()
            let repairCost = await estimateCost(request.judgeModel.provider, request.judgeModel.modelId, repairUsage)

        if let analysis = FusionJudgeParser.parse(repairResponse.text) {
            return JudgePhaseResult(
                    analysis: analysis,
                    rawJSON: repairResponse.text,
                    parseSucceeded: true,
                    latencyMs: repairLatency,
                    inputTokens: repairUsage?.inputTokens,
                    outputTokens: repairUsage?.outputTokens,
                    cost: (cost ?? 0) + (repairCost ?? 0),
                    error: nil,
                    piExecutions: [response.piExecution, repairResponse.piExecution].compactMap { $0 }
                )
        }
        throw ProviderClientError.parseFailure("Judge JSON parse failed")
    }

    private static func sanitizedModelId(_ modelId: String) -> String {
        let trimmed = modelId.trimmingCharacters(in: .whitespacesAndNewlines)
        let replaced = trimmed
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: ":", with: "_")
            .replacingOccurrences(of: " ", with: "_")
        return replaced.isEmpty ? "unknown" : replaced
    }

    private static func sumCosts(
        panelResults: [PanelResult],
        judge: JudgePhaseResult?,
        synthesis: SynthesisPhaseResult?
    ) -> Double? {
        var total: Double = 0
        var hasCost = false
        for panel in panelResults {
            if let cost = panel.cost {
                total += cost
                hasCost = true
            }
        }
        if let cost = judge?.cost {
            total += cost
            hasCost = true
        }
        if let cost = synthesis?.cost {
            total += cost
            hasCost = true
        }
        return hasCost ? total : nil
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/fusion/FusionOrchestrator.kt
package com.porarri.yamabikochat.data.fusion

import com.porarri.yamabikochat.data.model.ProviderRequest
import com.porarri.yamabikochat.data.model.ProviderRequestMessage
import com.porarri.yamabikochat.data.model.ProviderResponse
import com.porarri.yamabikochat.data.model.ProviderUsage
import java.util.UUID

typealias FusionOrchestratorInvoke =
    suspend (FusionPanelRunner.GenerateRequestBundle, FusionPhase, (com.porarri.yamabikochat.data.local.ToolActivityPayload) -> Unit) -> ProviderResponse
typealias FusionOrchestratorCostEstimator =
    suspend (provider: String, model: String, usage: ProviderUsage?) -> Double?
typealias FusionOrchestratorRequestBuilder = suspend (
    model: PanelModelConfig,
    systemPrompt: String,
    phase: FusionPhase,
    allowTools: Boolean,
    maxTokens: Int
) -> ProviderRequest
typealias FusionOrchestratorProgressHandler = (FusionProgressSnapshot) -> Unit

class FusionOrchestrator {
    suspend fun runThroughJudge(
        request: FusionRequest,
        context: FusionContext,
        buildRequest: FusionOrchestratorRequestBuilder,
        invoke: FusionOrchestratorInvoke,
        estimateCost: FusionOrchestratorCostEstimator,
        onProgress: FusionOrchestratorProgressHandler? = null
    ): FusionJudgeOutcome {
        val requestId = UUID.randomUUID().toString()
        val startedAtMs = System.currentTimeMillis()

        if (context.fusionDepth >= FusionContext.MAX_FUSION_DEPTH) {
            return runRecursionFallback(
                request = request,
                context = context,
                requestId = requestId,
                startedAtMs = startedAtMs,
                buildRequest = buildRequest,
                invoke = invoke,
                estimateCost = estimateCost
            )
        }

        val panelResults = FusionPanelRunner.runAll(
            request = request,
            panelSystemPrompt = FusionPrompts.panelSystemPrompt(request.taskType),
            buildPanelRequest = { panel, systemPrompt ->
                val genRequest = buildRequest(
                    panel,
                    systemPrompt,
                    FusionPhase.panel,
                    request.allowWebSearch,
                    request.maxPanelTokens
                )
                FusionPanelRunner.GenerateRequestBundle(
                    model = panel.modelId,
                    provider = panel.provider.uppercase(),
                    request = genRequest
                )
            },
            invoke = invoke,
            estimateCost = estimateCost,
            onProgress = onProgress
        )

        val successfulPanels = panelResults.filter { it.success }
        val failedModels = panelResults.filter { !it.success }.map { it.modelId }

        if (successfulPanels.isEmpty()) {
            throw FusionError.AllPanelsFailed(panelResults)
        }

        val finalPanelChips = panelResults.map { result ->
            FusionPanelChipStatus(
                modelId = result.modelId,
                provider = result.provider,
                state = if (result.success) FusionPanelChipState.succeeded else FusionPanelChipState.failed
            )
        }
        onProgress?.invoke(FusionProgressSnapshot.phaseOnly(FusionPhase.judge, finalPanelChips))

        val judgeOutcome = runJudge(
            request = request,
            successfulPanels = successfulPanels,
            failedModels = failedModels,
            buildRequest = buildRequest,
            invoke = invoke,
            estimateCost = estimateCost
        )

        val judgeAnalysis = judgeOutcome.analysis
        val synthesisUserPrompt = if (judgeAnalysis != null) {
            FusionPrompts.synthesizerUserPrompt(
                userPrompt = request.userPrompt,
                judgeAnalysis = judgeAnalysis,
                rawPanels = successfulPanels,
                judgeParseFailed = !judgeOutcome.parseSucceeded
            )
        } else {
            FusionPrompts.synthesizerUserPromptWithoutJudge(
                userPrompt = request.userPrompt,
                rawPanels = successfulPanels
            )
        }

        val synthSystemPrompt = FusionPrompts.synthesizerSystemPrompt(debugMode = context.debugMode)
        val mergedSystem = mergeSystemPrompts(request.systemPrompt, synthSystemPrompt)
        val synthesisRequest = buildRequest(
            request.synthesizerModel,
            synthSystemPrompt,
            FusionPhase.synthesizer,
            false,
            4096
        ).copy(
            messages = listOf(
                ProviderRequestMessage(role = "user", content = synthesisUserPrompt)
            ),
            systemPrompt = mergedSystem
        )

        val staticFallback = buildStaticFallback(
            judgeAnalysis = judgeAnalysis,
            judgeRaw = judgeOutcome.rawJSON,
            panels = successfulPanels
        )

        val trace = FusionTrace(
            requestId = requestId,
            preset = request.preset,
            startedAtMs = startedAtMs,
            completedAtMs = null,
            panelResults = panelResults,
            judgeResult = judgeOutcome,
            synthesisResult = null,
            totalLatencyMs = null,
            totalCost = sumCosts(panelResults = panelResults, judge = judgeOutcome, synthesis = null),
            failedModels = failedModels,
            status = "judge_complete",
            userPrompt = if (context.logPrompts) request.userPrompt else null,
            finalAnswer = null
        )

        val panelUsages = panelResults.mapNotNull { panel ->
            if (!panel.success) return@mapNotNull null
            val usage = tokenUsageOrNull(panel.inputTokens, panel.outputTokens)
            val requestType = "fusion_panel_${sanitizedModelId(panel.modelId)}"
            FusionTokenUsageRecord(panel.provider, panel.modelId, usage, requestType)
        }

        val judgeUsage = if (judgeOutcome.parseSucceeded || judgeOutcome.rawJSON != null) {
            FusionTokenUsageRecord(
                provider = request.judgeModel.provider.uppercase(),
                model = request.judgeModel.modelId,
                usage = tokenUsageOrNull(judgeOutcome.inputTokens, judgeOutcome.outputTokens),
                requestType = "fusion_judge"
            )
        } else {
            null
        }

        onProgress?.invoke(FusionProgressSnapshot.phaseOnly(FusionPhase.synthesizer, finalPanelChips))

        return FusionJudgeOutcome(
            trace = trace,
            synthesisRequest = synthesisRequest,
            synthesizerProvider = request.synthesizerModel.provider.uppercase(),
            synthesizerModel = request.synthesizerModel,
            staticFallbackAnswer = staticFallback,
            panelTokenUsages = panelUsages,
            judgeTokenUsage = judgeUsage
        )
    }

    fun finalizeTrace(
        trace: FusionTrace,
        synthesisResult: SynthesisPhaseResult,
        finalAnswer: String,
        logPrompts: Boolean
    ): FusionTrace {
        val completedAtMs = System.currentTimeMillis()
        return trace.copy(
            synthesisResult = synthesisResult,
            completedAtMs = completedAtMs,
            totalLatencyMs = completedAtMs - trace.startedAtMs,
            totalCost = sumCosts(
                panelResults = trace.panelResults,
                judge = trace.judgeResult,
                synthesis = synthesisResult
            ),
            status = if (synthesisResult.success) "completed" else "synthesis_fallback",
            finalAnswer = if (logPrompts) finalAnswer else null
        )
    }

    fun buildStaticFallback(
        judgeAnalysis: JudgeAnalysis?,
        judgeRaw: String?,
        panels: List<PanelResult>
    ): String = Companion.buildStaticFallback(judgeAnalysis, judgeRaw, panels)

    private suspend fun runJudge(
        request: FusionRequest,
        successfulPanels: List<PanelResult>,
        failedModels: List<String>,
        buildRequest: FusionOrchestratorRequestBuilder,
        invoke: FusionOrchestratorInvoke,
        estimateCost: FusionOrchestratorCostEstimator
    ): JudgePhaseResult {
        val started = System.currentTimeMillis()
        val judgeUserContent = FusionPrompts.judgeUserPrompt(
            userPrompt = request.userPrompt,
            successfulPanels = successfulPanels,
            failedModels = failedModels
        )

        suspend fun makeJudgeRequest(systemPrompt: String, userContent: String): FusionPanelRunner.GenerateRequestBundle {
            val base = buildRequest(
                request.judgeModel,
                systemPrompt,
                FusionPhase.judge,
                false,
                4096
            )
            val provider = request.judgeModel.provider.uppercase()
            return FusionPanelRunner.GenerateRequestBundle(
                model = request.judgeModel.modelId,
                provider = provider,
                request = base.copy(
                    messages = listOf(
                        ProviderRequestMessage(role = "user", content = userContent)
                    )
                )
            )
        }

        return try {
            val response = invoke(
                makeJudgeRequest(FusionPrompts.judgeSystemPrompt(), judgeUserContent),
                FusionPhase.judge,
                {}
            )
            val latencyMs = System.currentTimeMillis() - started
            val cost = estimateCost(
                request.judgeModel.provider,
                request.judgeModel.modelId,
                response.usage
            )

            val parsed = FusionJudgeParser.parse(response.text)
            if (parsed != null) {
                return JudgePhaseResult(
                    analysis = parsed,
                    rawJSON = response.text,
                    parseSucceeded = true,
                    latencyMs = latencyMs,
                    inputTokens = response.usage?.inputTokens,
                    outputTokens = response.usage?.outputTokens,
                    cost = cost,
                    error = null
                )
            }

            val repairStarted = System.currentTimeMillis()
            val repairResponse = invoke(
                makeJudgeRequest(
                    FusionPrompts.judgeSystemPrompt(),
                    FusionPrompts.jsonRepairPrompt(invalidJSON = response.text)
                ),
                FusionPhase.judge,
                {}
            )
            val repairLatency = System.currentTimeMillis() - repairStarted
            val repairCost = estimateCost(
                request.judgeModel.provider,
                request.judgeModel.modelId,
                repairResponse.usage
            )
            val repaired = FusionJudgeParser.parse(repairResponse.text)
            if (repaired != null) {
                return JudgePhaseResult(
                    analysis = repaired,
                    rawJSON = repairResponse.text,
                    parseSucceeded = true,
                    latencyMs = repairLatency,
                    inputTokens = repairResponse.usage?.inputTokens,
                    outputTokens = repairResponse.usage?.outputTokens,
                    cost = (cost ?: 0.0) + (repairCost ?: 0.0),
                    error = null
                )
            }

            JudgePhaseResult(
                analysis = null,
                rawJSON = repairResponse.text,
                parseSucceeded = false,
                latencyMs = repairLatency,
                inputTokens = repairResponse.usage?.inputTokens,
                outputTokens = repairResponse.usage?.outputTokens,
                cost = (cost ?: 0.0) + (repairCost ?: 0.0),
                error = "Judge JSON parse failed"
            )
        } catch (e: Exception) {
            JudgePhaseResult(
                analysis = null,
                rawJSON = null,
                parseSucceeded = false,
                latencyMs = System.currentTimeMillis() - started,
                inputTokens = null,
                outputTokens = null,
                cost = null,
                error = e.message ?: e.toString()
            )
        }
    }

    private suspend fun runRecursionFallback(
        request: FusionRequest,
        context: FusionContext,
        requestId: String,
        startedAtMs: Long,
        buildRequest: FusionOrchestratorRequestBuilder,
        invoke: FusionOrchestratorInvoke,
        estimateCost: FusionOrchestratorCostEstimator
    ): FusionJudgeOutcome {
        val fallback = request.synthesizerModel
        val genRequest = buildRequest(
            fallback,
            request.systemPrompt.orEmpty(),
            FusionPhase.fallback,
            false,
            4096
        ).copy(
            messages = listOf(
                ProviderRequestMessage(role = "user", content = request.userPrompt)
            )
        )
        val bundle = FusionPanelRunner.GenerateRequestBundle(
            model = fallback.modelId,
            provider = fallback.provider.uppercase(),
            request = genRequest
        )
        val started = System.currentTimeMillis()
        val response = invoke(bundle, FusionPhase.fallback) {}
        val latencyMs = System.currentTimeMillis() - started
        val cost = estimateCost(fallback.provider, fallback.modelId, response.usage)

        val synthesisResult = SynthesisPhaseResult(
            modelId = fallback.modelId,
            provider = fallback.provider.uppercase(),
            success = true,
            content = response.text,
            latencyMs = latencyMs,
            inputTokens = response.usage?.inputTokens,
            outputTokens = response.usage?.outputTokens,
            cost = cost,
            error = null,
            usedFallback = true
        )
        val trace = FusionTrace(
            requestId = requestId,
            preset = request.preset,
            startedAtMs = startedAtMs,
            completedAtMs = System.currentTimeMillis(),
            panelResults = emptyList(),
            judgeResult = null,
            synthesisResult = synthesisResult,
            totalLatencyMs = latencyMs,
            totalCost = cost,
            failedModels = emptyList(),
            status = "recursion_fallback",
            userPrompt = if (context.logPrompts) request.userPrompt else null,
            finalAnswer = if (context.logPrompts) response.text else null
        )

        return FusionJudgeOutcome(
            trace = trace,
            synthesisRequest = genRequest,
            synthesizerProvider = fallback.provider.uppercase(),
            synthesizerModel = fallback,
            staticFallbackAnswer = response.text,
            panelTokenUsages = emptyList(),
            judgeTokenUsage = null
        )
    }

    companion object {
        fun buildStaticFallback(
            judgeAnalysis: JudgeAnalysis?,
            judgeRaw: String?,
            panels: List<PanelResult>
        ): String {
            val sections = mutableListOf<String>()
            if (judgeAnalysis != null) {
                when {
                    judgeAnalysis.recommendedFinalPosition.isNotEmpty() ->
                        sections.add(judgeAnalysis.recommendedFinalPosition)
                    judgeAnalysis.consensus.isNotEmpty() ->
                        sections.add(judgeAnalysis.consensus.joinToString("\n"))
                }
            } else if (!judgeRaw.isNullOrEmpty()) {
                sections.add(judgeRaw)
            }
            val best = panels.filter { it.success }.minByOrNull { it.latencyMs }
            if (best != null && best.content.isNotEmpty()) {
                sections.add(best.content)
            }
            return sections.joinToString("\n\n").trim()
        }

        fun sanitizedModelId(modelId: String): String {
            val trimmed = modelId.trim()
            val replaced = trimmed
                .replace("/", "_")
                .replace(":", "_")
                .replace(" ", "_")
            return replaced.ifEmpty { "unknown" }
        }

        fun sumCosts(
            panelResults: List<PanelResult>,
            judge: JudgePhaseResult?,
            synthesis: SynthesisPhaseResult?
        ): Double? {
            var total = 0.0
            var hasCost = false
            for (panel in panelResults) {
                val cost = panel.cost
                if (cost != null) {
                    total += cost
                    hasCost = true
                }
            }
            judge?.cost?.let {
                total += it
                hasCost = true
            }
            synthesis?.cost?.let {
                total += it
                hasCost = true
            }
            return if (hasCost) total else null
        }

        fun mergeSystemPrompts(conversation: String?, fusion: String): String? {
            val left = conversation?.trim().orEmpty()
            val right = fusion.trim()
            return when {
                left.isEmpty() && right.isEmpty() -> null
                left.isEmpty() -> right
                right.isEmpty() -> left
                else -> "$left\n\n$right"
            }
        }

        fun tokenUsageOrNull(input: Int?, output: Int?): ProviderUsage? {
            if (input == null && output == null) return null
            return ProviderUsage(
                inputTokens = input ?: 0,
                outputTokens = output ?: 0,
                totalTokens = (input ?: 0) + (output ?: 0)
            ).normalizedNonEmpty()
        }
    }
}
